main.tf

```diff
@@ -1,3 +1,135 @@
+
+  # Determine what KMS service is being used for encryption
+  kms_service = var.existing_kms_instance_crn != null ? (
+    can(regex(".*kms.*", var.existing_kms_instance_crn)) ? "kms" : (
+      can(regex(".*hs-crypto.*", var.existing_kms_instance_crn)) ? "hs-crypto" : null
+    )
+  ) : null
+}
+
+resource "ibm_resource_instance" "en_instance" {
+  plan              = var.plan
+  location          = var.region
+  name              = var.name
+  resource_group_id = var.resource_group_id
+  tags              = var.tags
+  service           = "event-notifications"
+
+  parameters = {
+    service-endpoints = var.service_endpoints
+  }
+}
+
+#############################################################################
+# Event Notification KMS integration
+#############################################################################
+
+locals {
+  en_integration_id = [
+    for integrations in data.ibm_en_integrations.en_integrations.integrations :
+    integrations.id if integrations.type == local.kms_service
+  ]
+}
+data "ibm_en_integrations" "en_integrations" {
+  instance_guid = ibm_resource_instance.en_instance.guid
+}
+
+resource "ibm_en_integration" "en_kms_integration" {
+  count          = var.kms_encryption_enabled == false || var.skip_iam_authorization_policy ? 0 : 1
+  instance_guid  = ibm_resource_instance.en_instance.guid
+  integration_id = local.en_integration_id[0]
+  type           = local.kms_service
+  metadata {
+    endpoint    = var.kms_endpoint == "public" ? "https://${var.kms_region}.kms.cloud.ibm.com" : "https://private.${var.kms_region}.kms.cloud.ibm.com"
+    crn         = var.existing_kms_instance_crn
+    root_key_id = var.root_key_id
+  }
+}
+
+##############################################################################
+# IAM Authorization Policy
+##############################################################################
+
+locals {
+  existing_kms_instance_guid = element(split(":", var.existing_kms_instance_crn), length(split(":", var.existing_kms_instance_crn)) - 3)
+}
+
+# Create IAM Authorization Policies to allow event notification to access kms for the encryption key
+resource "ibm_iam_authorization_policy" "kms_policy" {
+  count                       = var.kms_encryption_enabled == false || var.skip_iam_authorization_policy ? 0 : 1
+  source_service_name         = "event-notifications"
+  source_resource_instance_id = ibm_resource_instance.en_instance.guid
+  target_service_name         = local.kms_service
+  target_resource_instance_id = local.existing_kms_instance_guid
+  roles                       = ["Reader"]
+  description                 = "Allow all Event Notification instances in the resource group ${var.resource_group_id} to read from the ${local.kms_service} instance GUID ${var.existing_kms_instance_crn}"
```

Comment:
This description is not accurate since you have scoped it down to use the `source_resource_instance_id`. If this is working (which I assume you have verified?) then suggest updating to this:

"Allow Event Notification instance ${ibm_resource_instance.en_instance.guid} to read from the ${local.kms_service} instance ${local.existing_kms_instance_guid}"

Reply:
Yes I tested,
<img width="1396" alt="image" src="https://github.com/terraform-ibm-modules/terraform-ibm-event-notifications/assets/4889052/5fd1beee-222f-4abb-9b71-26ce51532dba">

Update the description also.